path_rotate_test.go

```diff
@@ -198,3 +204,88 @@ func TestManualRotate(t *testing.T) {
 		}
 	})
 }
+
+type failingRollbackClient struct {
+	count    int
+	maxCount int
+	password string
+}
+
+func (f *failingRollbackClient) UpdateDNPassword(conf *client.Config, dn string, newPassword string) error {
+	f.count += 1
+	if f.count >= f.maxCount {
+		f.password = newPassword
+		return nil
+	}
+	return fmt.Errorf("some error")
+}
+
+func (f *failingRollbackClient) UpdateUserPassword(conf *client.Config, user, newPassword string) error {
+	panic("nope")
+}
+
+func (f *failingRollbackClient) Execute(conf *client.Config, entries []*ldif.Entry, continueOnError bool) error {
+	panic("nope")
+}
+
+var _ ldapClient = (*failingRollbackClient)(nil)
+
+func TestRollback(t *testing.T) {
+	oldRollbackAttempts, oldMinRollbackDuration, oldMaxRollbackDuration := rollbackAttempts, minRollbackDuration, maxRollbackDuration
+	t.Cleanup(func() {
+		rollbackAttempts = oldRollbackAttempts
+		minRollbackDuration = oldMinRollbackDuration
+		maxRollbackDuration = oldMaxRollbackDuration
+	})
+	rollbackAttempts = 5
+	minRollbackDuration = 1 * time.Millisecond
+	maxRollbackDuration = 10 * time.Millisecond
+
+	fclient := &failingRollbackClient{}
+	b := &backend{
+		client: fclient,
+	}
+	cfg := &config{
+		LDAP: &client.Config{
+			ConfigEntry: &ldaputil.ConfigEntry{
+				BindDN: "",
+			},
+		},
+	}
+
+	ctx := context.Background()
+
+	// works if the client always succeeds
```

Comment:
I love the tests here! I think converting this to be a table driven test might be a cleaner approach. We could get a fake client for each sub test.